imjoy/helper.py

```diff
@@ -62,46 +64,155 @@ def killProcess(logger, pid):
+        )
+
+    def _process_start(pid=None, cmd=None):
+        """Run after starting the process."""
+        logging_callback(f"Running requirements subprocess(pid={pid}): {cmd}")
+        process_start(pid=pid)
+
+    logger.info("Running requirements commands: %s", commands)
+    code, errors = run_commands(env, work_dir, commands, _process_start, process_finish)
+
+    if code == 0:
+        cmd_history.extend(commands)
+        logging_callback("Requirements command executed successfully.")
+        return
+    logging_callback(f"Failed to run requirements command: {commands}", type="error")
+    if errors is not None:
+        logging_callback(str(errors, "utf-8"), type="error")
+
+    if not eng.opt.CONDA_AVAILABLE:
+        fail_install()
+
+    git_cmd = ""
+    if shutil.which("git") is None:
+        git_cmd += " git"
+    if shutil.which("pip") is None:
+        git_cmd += " pip"
+    if git_cmd == "":
+        fail_install()
+
+    logger.info("pip command failed, trying to install git and pip")
+    # try to install git and pip
+    git_cmd = f"conda install -y{git_cmd}"
+    code, _ = run_process(
+        git_cmd.split(),
+        process_start=_process_start,
+        stderr=None,
+        env=env,
+        cwd=work_dir,
+    )
+    if code != 0:
+        fail_install()
+
+    logger.info("Running requirements commands: %s", commands)
+    code, errors = run_commands(env, work_dir, commands, _process_start, process_finish)
+
+    if code != 0:
+        fail_install()
+
+
+def run_commands(env, work_dir, commands, process_start, process_finish):
+    """Run commands in separate processes."""
+    code = 0
+    errors = None
+
+    for cmd in commands:
+        print("Running command:", cmd)
+        code, errors = run_process(
+            cmd,
+            process_start=process_start,
+            process_finish=process_finish,
+            shell=True,
+            env=env,
+            cwd=work_dir,
+        )
+        if code:
+            return code, errors
+
+    return code, errors
+
+
+def run_process(cmd, process_start=None, process_finish=None, **kwargs):
+    """Run subprocess command."""
+    shell = kwargs.pop("shell", False)
+    stderr = kwargs.pop("stderr", subprocess.PIPE)
+    process = subprocess.Popen(cmd, shell=shell, stderr=stderr, **kwargs)
+    if process_start is not None:
+        process_start(pid=process.pid, cmd=cmd)
+    return_code = process.wait()
+    if stderr is not None:
+        _, errors = process.communicate()
     else:
-        raise Exception("Unsupported requirements type.")
-    requirements_cmd = "{} && {}".format(requirements_cmd, default_command)
-    requirements_cmd = requirements_cmd.strip(" &")
-    return requirements_cmd
+        errors = None
+    if process_finish is not None:
+        process_finish(pid=process.pid, cmd=cmd)
+
+    return return_code, errors
 
 
 def parseEnv(eng, env, work_dir, default_env_name):
```

Comment:
The naming is not consistent with `pase_reqs`, I think it would be better to use full name, like that it will be easier for people to read. At a first glance it's hard to think `reqs` means requirements.